manta-accounting/src/transfer/mod.rs

```diff
@@ -2110,3 +2118,147 @@ where
+)]
+pub struct IdentityProof<C>
+where
+    C: Configuration,
+{
+    /// TransferPost
+    pub transfer_post: TransferPost<C>,
+}
+
+impl<C> IdentityProof<C>
+where
+    C: Configuration,
+{
+    /// Verifies `self` for `address` against `virtual_asset`.
+    ///
+    /// # Note
+    ///
+    /// It performs four checks:
+    ///
+    /// 1) `identity_proof.transfer_post` has a [`ToPublic`](crate::transfer::canonical::ToPublic)
+    /// shape.
+    ///
+    /// 2) `identity_proof.transfer_post` has a valid proof.
+    ///
+    /// 3) The [`UtxoAccumulatorOutput`] in `identity_proof.transfer_post` has been computed from
+    /// `virtual_asset` and `address`.
+    ///
+    /// 4) The virtual asset is correctly constructed.
+    #[inline]
+    pub fn identity_verification<A>(
+        &self,
+        parameters: &Parameters<C>,
+        verifying_context: &VerifyingContext<C>,
+        utxo_accumulator_model: &UtxoAccumulatorModel<C>,
+        virtual_asset: IdentifiedAsset<C>,
+        address: Address<C>,
+    ) -> Result<(), IdentityVerificationError>
+    where
+        C::UtxoAccumulatorOutput: PartialEq,
+        UtxoAccumulatorModel<C>: Clone + Model<Verification = bool>,
+        A: Accumulator<
+                Item = UtxoAccumulatorItem<C>,
+                Model = UtxoAccumulatorModel<C>,
+                Output = UtxoAccumulatorOutput<C>,
+            > + From<UtxoAccumulatorModel<C>>,
+        Asset<C>: Default,
+    {
+        let IdentifiedAsset::<C> { identifier, asset } = virtual_asset;
+        if <Parameters<C> as IdentifierTransparency>::is_transparent(&identifier)
+            || asset == Default::default()
+        {
+            return Err(IdentityVerificationError::InvalidIdentifiedAsset);
+        }
+        TransferShape::from_post(&self.transfer_post).map_or(
+            Err(IdentityVerificationError::InvalidShape),
+            |shape| match shape {
+                TransferShape::ToPublic => Ok(()),
+                _ => Err(IdentityVerificationError::InvalidShape),
+            },
+        )?;
+        if !self
+            .transfer_post
+            .has_valid_proof(verifying_context)
+            .map_err(|_| IdentityVerificationError::InvalidProof)?
+        {
+            return Err(IdentityVerificationError::InvalidProof);
+        }
+        let utxo = parameters.utxo_reconstruct(&asset, &identifier, &address);
+        let mut utxo_accumulator: A = utxo_accumulator_model.clone().into();
+        utxo_accumulator.insert(
+            &parameters
+                .utxo_accumulator_item_hash()
+                .item_hash(&utxo, &mut ()),
+        );
+        let utxo_accumulator_output = utxo_accumulator
+            .prove(
+                &parameters
+                    .utxo_accumulator_item_hash()
+                    .item_hash(&utxo, &mut ()),
+            )
+            .expect("Failed to generate UtxoMembershipProof")
```

Comment:
Because at this level, we don't know that our `Accumulator` is a Merkle forest, so we can't use the `root()` method to get the `UtxoAccumulatorOutput`. The only way (at least that I could find) to obtain an output using only `accumulator.rs` methods is through the `prove` function.

Reply:
Ok, then let's think of a way to extend the current APIs to get the root without having to build a proof that should be true for all accumulators. Perhaps something like

```rust
fn output_for(&self, item: &Self::Item) -> Option<Self::Output>;
```